src/tango_sdp_subarray/SDPSubarray/SDPSubarray.py

```diff
@@ -198,52 +251,76 @@ def ReleaseResources(self, config=''):
-        self._obs_state = ObsState.CONFIGURING
-        # time.sleep(1)
+        self._require_obs_state([ObsState.IDLE])
 
-        # Validate the SBI config schema
-        if schema_path is None:
-            schema_path = join(dirname(__file__), 'schema',
-                               'configure_pb.json')
+        LOG.info('Command: Configure (%s)', self.get_name())
+
+        LOG.debug('Setting ObsState to CONFIGURING.')
+        self._obs_state = ObsState.CONFIGURING
 
+        LOG.debug('Validating PB configuration.')
+        try:
+            pb_config = json.loads(pb_config)
+        except json.JSONDecodeError as error:
+            LOG.error('Unable to load JSON PB configuration: %s', error.msg)
+            raise
+        schema_path = join(dirname(__file__), 'schema', 'configure_pb.json')
         with open(schema_path, 'r') as file:
             schema = json.loads(file.read())
-
-        pb_config = json.loads(pb_config)
-        validate(pb_config, schema)
-
-        for txn in self.db_client.txn():
-            confdata = pb_config['configure']
-            pb = entity.ProcessingBlock(confdata['id'], None,
-                                        confdata['workflow'],
-                                        confdata['parameters'],
-                                        confdata['scanParameters'])
-            txn.create_processing_block(pb)
-
+        try:
+            validate(pb_config, schema)
+        except exceptions.ValidationError as error:
+            LOG.error('PB configuration failed to validate: %s', error.message)
+            frame_info = getframeinfo(currentframe())
+            error_message = '{} : {} : {}'.format(
+                str(error.absolute_path),
+                str(error.schema_path),
+                error.message
+            )
+            Except.throw_exception(
+                'Command: Configure failed', error_message,
+                '{}:{}'.format(frame_info.filename, frame_info.lineno))
+
+        # Add the PB configuration to the database.
+        if HAVE_CONFIG_DB and self._toggle_config_db:
+            for txn in self.db_client.txn():
+                conf_data = pb_config['configure']
+                pb = entity.ProcessingBlock(conf_data['id'],
+                                            None,
+                                            conf_data['workflow'],
+                                            conf_data['parameters'],
+                                            conf_data['scanParameters'])
+                txn.create_processing_block(pb)
+
+        # Cache a local copy of the PB configuration.
+        self._pb_config = pb_config['configure']
+
+        # Evaluate the receive hosts <-> channel mapping.
+        self._evaluate_receive_host_channel_map()
+
+        # Evaluate the receive addresses.
+        # NOTE(BMo) Depending on the deployment model this may need to be \
+        # delayed until EE containers are started.
```

Comment:
And the channel link map is available...

Reply:
Yes very true. I'll Update the comment to make sure both qualifications are noted. I was implicitly assuming the channel link map part of that because the code that comment is linked currently reads the channel link map (`cbfOutputLink` attribute) before generating the `receiveAddresses` dictionary.

There is an interesting concurrency question still to be fully resolved however here as we can't just read the `cbfOutputLink` attribute at any point in the configure. Instead we need to "wait" on the attribute push change event to know that it has been updated. To do that we can either use a blocking wait or a event subscription with a callback function. The next version of this code (which i'm testing now) should address this.